### ISSUE TITLE
this looks interesting but...

### ISSUE THREAD

**Issue by rednaxus** [valet-io/bookshelf-authorization#1]
does it actually work as described?  Is this a draft in progress?

**Comment by bendrucker** [comment]
Most certainly works! Check out the integration test folder for a real world example: https://github.com/valet-io/bookshelf-authorization/tree/master/test/integration

I've yet to upgrade for compatibility with Bookshelf 0.7, but most if not all of the API should be compatible.

**Comment by rednaxus** [comment]
Ok, awesome... I like the approach...

the unit tests you have don't seem to work though... and i can't seem to figure out how to use this

For example:

```
    it('can read when there are 1+ appointments', function(){
      patient.appointments = [{doctor: doctor}];
      return doctor.can.read(patient);
   });
```

but as far as I can tell, doctor.can.read is a Promise?  so this case it always seems true, no matter how I have set the privileges

**Comment by rednaxus** [comment]
ok, some more information... your package.json has bluebird ~1.0.6 which is causing an extra 'throw' of the exception... when I switch to ~1.2.0 everything is ok....

except the example above still needs to be something like

```
    doctor.can.read(patient).then(function (){ ..success...})
    .catch(function(err){ ...error....})
```

**Comment by bendrucker** [comment]
> as far as I can tell, doctor.can.read is a Promise

Yes, it must be or else you can't do async authorization. And it definitely is not always true. Look at the test below where the promise is rejected and read the authorization function set up in the relevant beforeEach. 

No idea what you mean by an extra throw. 100% of the tests are passing so I'm not sure what you're doing wrong. Mocha test functions can return promises and if they resolve the test passes.

I'll certainly upgrade the deps and release for Bookshelf 0.7, but this is fully functional and in active use now.

**Comment by rednaxus** [comment]
ok thanks.... so here is what i get with your standard package installation:

Possibly unhandled AuthorizationError: 
  at [object Object].<anonymous> (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/lib/rule.js:26:11)
  at tryCatch2 (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/util.js:74:19)
  at [object Object].Promise$_method [as run](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/promise.js:292:25)
  at [object Object].<anonymous> (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/lib/can.js:31:19)
  at Object.Promise$_mapper (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/map.js:63:32)
  at tryCatch1 (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/util.js:64:19)
  at Promise$_callHandler [as _callHandler](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/promise.js:708:13)
  at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/promise.js:724:18)
  at Promise$_settlePromiseAt [as _settlePromiseAt](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/promise.js:896:14)
  at Promise$_fulfillPromises [as _fulfillPromises](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/promise.js:1041:14)
  at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer](/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/async.js:64:12)
  at Async$consumeFunctionBuffer (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/async.js:37:14)
  at Promise$_Scheduler (/home/alan/dev/workspace/entepath/s1/node_modules/bookshelf-authorization/node_modules/bluebird/js/main/schedule.js:40:58)
  at process._tickDomainCallback (node.js:463:13)

the test code (jasmine-node):

```
    doctor.can.read(patient).then (res)->
      expect(true).toBe false
      done()
    .catch (err)->
      expect(err).toBe err
      done()
```

In this case, doctor has no read rights so this causes an err to be generated and a catch, which works fine, but bluebird seems to throw that additional unhandled exception..

This is with bluebird 1.0.8

If I change your package.json to use bluebird ~1.2.0 and npm install it (getting 1.2.4), then the 'extra' exception reported above doesn't get thrown, everything works as expected....  Very repeatable, and on different platforms

ok, some more data:
if I explicitly install bluebird 1.0.5, then everything is ok!  This is probably why you don't see it.

So most likely it's a bug in 1.0.8...

**Comment by bendrucker** [comment]
It's not per se re-throwing an exception. Bluebird's [cleverness](https://github.com/petkaantonov/bluebird#error-handling) with possiblyUnhandled is that it waits a full tick to allow exceptions to be caught before logging them. Most promise implementations just swallow uncaught exceptions with no indication of failure. There have definitely been bugs in the past with this behavior since it's tricky to get right. I was looking @ Travis when I said that I hadn't seen anything. A regression introduced in a minor version would explain things. 

I've added a few things to target a v0.1. Hoping I have time to bang them out this weekend. 

Closing this, feel free to open additional issues if there are things you'd like prioritized for 0.1 or wishes for the future. 

https://github.com/valet-io/bookshelf-authorization/issues?milestone=1&state=open
